modules/python/python-options.c

```diff
@@ -0,0 +1,336 @@
+  self->value = value;
+
+  return &self->super;
+}
+
+/* String List */
+
+typedef struct _PythonOptionStringList
+{
+  PythonOption super;
+  GList *value;
+} PythonOptionStringList;
+
+static PyObject *
+_string_list_create_value_py_object(const PythonOption *s)
+{
+  PythonOptionStringList *self = (PythonOptionStringList *) s;
+  return py_string_list_from_string_list(self->value);
+}
+
+static void
+_string_list_free_fn(PythonOption *s)
+{
+  PythonOptionStringList *self = (PythonOptionStringList *) s;
+  string_list_free(self->value);
+}
+
+PythonOption *
+python_option_string_list_new(const gchar *name, const GList *value)
+{
+  PythonOptionStringList *self = g_new0(PythonOptionStringList, 1);
+  python_option_init_instance(&self->super, name);
+
+  self->super.create_value_py_object = _string_list_create_value_py_object;
+  self->super.free_fn = _string_list_free_fn;
+  self->value = string_list_clone(value);
+
+  return &self->super;
+}
+
+/* Template */
+
+typedef struct _PythonOptionTemplate
+{
+  PythonOption super;
+  gchar *value;
+} PythonOptionTemplate;
+
+static PyObject *
+_template_create_value_py_object(const PythonOption *s)
+{
+  PythonOptionTemplate *self = (PythonOptionTemplate *) s;
+
+  PyObject *template_str = py_string_from_string(self->value, -1);
+  if (!template_str)
+    return NULL;
+
+  PyObject *args = PyTuple_Pack(1, template_str);
+  PyObject *py_template = PyObject_Call((PyObject *) &py_log_template_type, args, NULL);
+
+  Py_DECREF(template_str);
+  Py_DECREF(args);
+  return py_template;
+}
+
+static void
+_template_free_fn(PythonOption *s)
+{
+  PythonOptionTemplate *self = (PythonOptionTemplate *) s;
+  g_free(self->value);
+}
+
+PythonOption *
+python_option_template_new(const gchar *name, const gchar *value)
+{
+  PythonOptionTemplate *self = g_new0(PythonOptionTemplate, 1);
+  python_option_init_instance(&self->super, name);
+
+  self->super.create_value_py_object = _template_create_value_py_object;
+  self->super.free_fn = _template_free_fn;
+  self->value = g_strdup(value);
+
+  return &self->super;
+}
+
+/* Python Options */
+
+struct _PythonOptions
+{
+  GList *options;
+};
+
+PythonOptions *
+python_options_new(void)
+{
+  PythonOptions *self = g_new0(PythonOptions, 1);
+  return self;
+}
+
+void
+python_options_add_option(PythonOptions *self, PythonOption *option)
+{
+  self->options = g_list_append(self->options, option);
+}
+
+PyObject *
+python_options_create_py_dict(const PythonOptions *self)
+{
+  PyObject *py_dict;
+
+  PyGILState_STATE gstate = PyGILState_Ensure();
+  {
+    py_dict = PyDict_New();
+    for (GList *elem = self->options; elem; elem = elem->next)
+      {
+        const PythonOption *option = (const PythonOption *) elem->data;
+        const gchar *name = python_option_get_name(option);
+        PyObject *value = python_option_create_value_py_object(option);
+
+        if (!value)
+          continue;
+
+        PyDict_SetItemString(py_dict, name, value);
```

Comment:
This and the constructor call may fail. I don't know how much this actually happens. Probably in OOM cases.

Reply:
Thanks, I have handled those errors. In case of ctor fail, we can only return NULL, in case of set failure I have added an error message and continued.